sushant94-dev PR

sushant94 opened the pull request:
Not for merge. There is still work to be done before this PR is ready to me merged with master.

**EDIT (19th Oct):** TODO Before merge:
- [x] #100 
- [x] #101 
- [x] #102 
- [x] #103 

**Edit (27th Oct):**
- [ ] Need to decide which trace statements to keep inside phipalcement.rs, there seem to be too many, but super helpful in debugging.

sushant94 committed:
Uses newly added OpRol and OpRor

sushant94 committed:
Implements more reference marking

sushant94 committed:
Adds reference data to text IR

sushant94 committed:
Adds initial support for custom opcodes

sushant94 committed:
Implements InstructionAnalyzer

sushant94 committed:
Implements handling of unhandled esil instructions

sushant94 committed:
Fixes panic in phiplacement.rs

sushant94 committed:
Implements more containers framework

sushant94 committed:
Implements more containers

XVilka commented on minidec/main.rs:
```diff
@@ -94,7 +97,7 @@ fn main() {
         {
             println!("  [*] Eliminating Dead Code");
-            dce::collect(&mut rfn.ssa);
+            //dce::collect(&mut rfn.ssa);
```
You disabled DCE?

XVilka commented on src/frontend/instruction_analyzer.rs:
```diff
@@ -0,0 +1,193 @@
+    }
+}
+
+/// Capstone-based x86 instruction analyzer
```
But WHY?

XVilka commented on src/frontend/instruction_analyzer.rs:
```diff
@@ -0,0 +1,193 @@
+                                IOperand::Invalid
+                            }
+                        },
+                        cs::x86_op_type::X86_OP_MEM => {
```
This information is available in `ao[j]` output. If you need detailed information - there is an 'opex' field.

sushant94 committed:
Starts work on `radeco_source`, replacement to `source`

- Better error handling in `Source`
- R2 no longer needs to be mutable, this means we can
  finally pass around the r2struct without worrying about it
  (See implementation of `WrappedR2Api<R>`

sushant94 committed:
Completes Version1 of RadecoContainers

chinmaydd commented on src/frontend/radeco_containers.rs:
```diff
@@ -110,47 +161,272 @@ pub enum FunctionType {
+        self
+    }
+
+    /// Set the source to use for loading. This is propagated to every `ModuleLoader`
```
Why do we need to replicate the entire Source for each ModuleLoader? Naive doubt, but doesnt it increase redundancy?

sushant94 replied:
This allows all the modules loaders to work independently to load themselves up (using source). It also allows us to load different modules from different sources.

For example, you could load the main binary and perform analysis using r2 (-source), but for the libraries (which don't change regardless of which binary they're used in) we could analyze them once, cache the analysis, and reload using the cached source.

From memory usage perspective, this isn't heavy, as all we're doing is storing a reference to the original source (as opposed to cloning it).

sushant94 commented on src/frontend/radeco_source.rs:
```diff
@@ -0,0 +1,315 @@
+}
+
+// TODO: Split this up/compose this from more basic traits to avoid reimplementation
+// as currently this is a re-implementation of r2api
```
Yes, so if you've looked at the r2pipe-api [repo](https://github.com/radare/radare2-r2pipe-api/blob/master/rust/src/api_trait.rs#L13), I've separated (using comments) the different functions in r2api based on the functionality it provides (such as loader, arch, low-level analysis etc). These information can come from separate sources too. So imo, it makes sense to split them up into separate traits and have the current R2Api trait as a composition of these primitive traits:
`pub trait R2Api: `Loader + `Arch` ... { }`

Additionally, the current source takes `self` immutably, while r2api takes `self` mutably. This needs to be discussed. Perhaps we can reduce redundancy by implementing r2api over `Rc<RefCell<R2>>` (as it is currently here), rather than over R2 directly.

XVilka replied:
I agree. Better to have the support of separate, independend sources.

sushant94 committed:
Improves docs

sushant94 committed:
Loads more information into containers

- Adds loading of callgraph and datarefs
- Closes #102

sushant94 committed:
Completes iterators. Closes #100.

sushant94 committed:
Adds deprecated warning to containers and source. Closes #101.

sushant94 committed:
Introduces constraint solver for mark refs

sushant94 committed:
Separates constraint solver logic

sushant94 committed:
Generalized `ConstraintSet` over type of bindings

sushant94 committed:
Clean working commit

ZhangZhuoSJTU commented:
Btw, @sushant94 may I ask what happened to Travis? I just restarted the last build on master branch, and it failed. (`capstone errors` for stable version, and `cargo test` failed for beta version)

sushant94 committed:
Fixes ton of panics in ssaconstruction. Needs cleanup.

sushant94 committed:
Continue work on reference marking

sushant94 committed:
Make logging and profiling optional features

sushant94 commented:
@ZhangZhuoSJTU That's ok. I'm planning to move away from using json files to using stable versions of r2 in tests, and bleeding-edge version of r2 as a tests which may fail. This will help us catch changes in r2 early, while not failing the tests in travis.

sushant94 committed:
Adds mutable iterators for radeco_containers

sushant94 committed:
Implements more reference marking

sushant94 committed:
Automatically create bindings in ConstraintSet

sushant94 committed:
Minor reorganization of code

sushant94 committed:
Implements initial version of interfunction analyzer

sushant94 committed:
Minor improvements to interproc analyzer

sushant94 committed:
More work on interproc reference marking

XVilka commented on Cargo.toml:
```diff
@@ -15,24 +15,35 @@ path = "minidec/main.rs"
 git = "https://github.com/radare/esil-rs"
+#path = "../esil-rs"
+
+[dependencies.capstone_rust]
```
Don't forget to kill this one.

XVilka commented on src/analysis/constraint_set.rs:
```diff
@@ -0,0 +1,349 @@
+//! Implements Constaint and ConstraintSet
```
"Constraint"

XVilka commented on src/frontend/containers.rs:
```diff
@@ -350,13 +351,13 @@ impl<'a, T: 'a + Source> From<&'a mut T> for RadecoModule<'a, DefaultFnTy> {
             let reg_info = reg_info.clone();
 
-            let handle = thread::spawn(move || {
+            let handle = thread::Builder::new().name(f.name.as_ref().unwrap().to_owned()).spawn(move || {
```
may be split this line?

sushant94 replied:
This doesn't matter. This whole file is set to be deprecated and we shouldn't be using it at all.

XVilka commented on src/frontend/ssaconstructor.rs:
```diff
@@ -503,6 +495,27 @@ impl<'a, T> SSAConstruct<'a, T>
+                    .map(|cap| u8::from_str_radix(&cap[0], 16).expect("Cannot Fail"))
+                    .collect::<Vec<u8>>();
+
+                // TODO/XXX:
```
And this should be done before merging too, I think.

XVilka commented on src/frontend/ssaconstructor.rs:
```diff
@@ -514,25 +527,168 @@ impl<'a, T> SSAConstruct<'a, T>
+            // Mulitply the index by the scale and return the resulting node
+            //    <index> '*' <scale>
+            let reg_node = self.process_in(&Some(Token::ERegister(reg.clone())), addr).expect("Invalid op");
+            // TODO: s/64/default op size/
```
Keep in mind ultra-long instructions too.

XVilka commented:
So far looks good. 👍

sushant94 committed:
Adds `VarBindings` and supporting structs

- Sets callgraph information in respective `RadecoFunction`s

sushant94 committed:
Nearly completes interprocedural reference marking

TODO:
    - Setup correct node index in `VariableBinding`
    - Update `CallContextInfo` in CallGraph

sushant94 committed:
Completes inter-procedural analysis for `ReferenceMarker`

    - Completes fixpoint check

XVilka commented:
Any update on this?

sushant94 commented:
Currently evaluating the implementation above against test cases. It'll take me a few days to get back to dev.

XVilka commented:
So?

XVilka commented:
Ping?

sushant94 committed:
Temporary commit

ZhangZhuoSJTU commented:
@sushant94 Ping. Would you mind fixing the travis error and merging this PR, which may help some future work basing on this new framework? But it's ok if merging takes some problems, for I could work on some other work ;P

sushant94 commented:
@ZhangZhuoSJTU I'm ready to merge this. But @XVilka wants the capstone dependency removed, and I currently cannot refactor that part to use radare2. I'll get back to working on this full time in a week.

XVilka commented:
.

sushant94 committed:
Adds block sizes to basic blocks

- TO FIX: Comments out tests that were broken.
- Some changes that are a part of this commit should not make it to master!

sushant94 committed:
Final commit before fix-tests and merge

sushant94 committed:
[deprecated] Removed tests from container.rs

sushant94 committed:
Adds tests in fixcall.rs to ignore

- Need to rewrite to use new containers and radare2 in tests

sushant94 committed:
Disables test in interproc.rs

sushant94 committed:
Cleans up debug output

sushant94 committed:
Adds radare2 to travis.yml

sushant94 committed:
Adds capstone to travis.yml

sushant94 committed:
Removes debug function
